packages/itwinui-react/src/core/FileUpload/FileUploadCard.tsx

```diff
@@ -0,0 +1,323 @@
+  const { data, onDataChange, setInternalData, inputId } = useSafeContext(
+    FileUploadCardContext,
+  );
+
+  const setNativeFilesRef = React.useCallback(
+    (node: HTMLInputElement | null) => {
+      if (!node || !getWindow()?.DataTransfer) {
+        return;
+      }
+
+      const dataTransfer = new DataTransfer();
+      dataTransfer.items.clear();
+      data.forEach((file) => dataTransfer.items.add(file));
+      node.files = dataTransfer.files;
+    },
+    [data],
+  );
+
+  const refs = useMergedRefs(ref, setNativeFilesRef);
+
+  return (
+    <>
+      <input
+        className={cx('iui-file-card-input', className)}
+        type='file'
+        onChange={(e) => {
+          onChange?.(e);
+          if (!e.isDefaultPrevented()) {
+            const _files = Array.from(e.currentTarget.files || []);
+            onDataChange?.(_files);
+            setInternalData(_files);
+          }
+        }}
+        ref={refs}
+        id={inputId}
+        {...rest}
+      />
+      {children}
+    </>
+  );
+});
+
+export type FileUploadCardProps = {
+  /**
+   * File to pass (only needed when passing a custom action)
+   */
+  data?: File[];
+  /**
+   * Callback fired when data has changed (only needed passing custom action)
+   */
+  onDataChange?: (data: File[]) => void;
+  /**
+   * Node that is shown when there is no file uploaded
+   * Either pass `FileEmptyCard` (for default state) or a different component to show
+   * @default <FileEmptyCard />
+   */
+  emptyCard?: React.ReactNode;
+  /**
+   * Input node
+   * @default <FileUploadCard.Input />
+   */
+  input?: React.ReactNode;
+} & React.ComponentPropsWithoutRef<'div'>;
+/**
+ * Default card to be used with the `FileUpload` wrapper component for single-file uploading.
+ * @example
+ * <FileUploadCard />
+ * <FileUploadCard data={data} onDataChange={(data) => setData(data)}>
+ *   <FileUploadCard.Icon>
+ *     <SvgSmileyHappyVery />
+ *   </FileUploadCard.Icon>
+ *   <FileUploadCard.Info>
+ *     <FileUploadCard.Title>Custom File Name</FileUploadCard.Title>
+ *     <FileUploadCard.Description>
+ *       Custom File Description
+ *     </FileUploadCard.Description>
+ *   </FileUploadCard.Info>
+ *   <FileUploadCard.Action>
+ *     <Button
+ *       onClick={() => {
+ *         setData([]);
+ *       }}
+ *     />
+ *     <FileUploadCard.Input name={fileInputId} ref={inputRef} />
+ *   </FileUploadCard.Action>
+ * </FileUploadCard>
+ */
+export const FileUploadCard = Object.assign(
+  React.forwardRef<HTMLDivElement, FileUploadCardProps>(
+    (props, ref: React.RefObject<HTMLDivElement>) => {
+      const {
+        className,
+        children,
+        data: dataProps,
+        onDataChange,
+        emptyCard = <FileEmptyCard />,
+        input,
+        ...rest
+      } = props;
+
+      const [internalData, setInternalData] = React.useState<File[]>();
+      const data = dataProps ?? internalData ?? [];
+      const inputId = React.useId();
```

Comment:
also i think you will need to check for user's own `id` and prefer that if passed.

e.g. the user can do this:
```tsx
<FileUploadCard input={<input id='blah'>} />
```

Reply:
> `useId` doesn't exist before react 18 and we aim to be compatible all the way back to react 16, so you will need to add a fallback.
> 
> you can create a utility like this:
> 
> ```js
> export const useId = React.useId ?? () => {
>   const [id] = React.useState(() => `iui-${getRandomValue(10)}`);
>   return id;
> };
> ```

This has been added. I am still working on the second part of this comment

Reply:
it should go in the hooks folder

Reply:
This has been moved

Reply:
> also i think you will need to check for user's own `id` and prefer that if passed.
> 
> e.g. the user can do this:
> 
> ```tsx
> <FileUploadCard input={<input id='blah'>} />
> ```

This has been changed, but I would like a second opinion on the implementation. I set the `inputId` using `useId()` first and then check for an `id` prop in `input` since `useId()` cannot be called conditionally.

Reply:
i don't think you can read `input.props` like that. you need to read it from inside props in `FileUploadCard.Input` and then set it in context.